### ISSUE TITLE
WebView Crash

### ISSUE THREAD

**Issue by bartutiryakioglu** [AppLovin/AppLovin-MAX-Unity-Plugin#122]
I am using only interstitial and rewarded ads and I came accross with a base.apk!libmonochrome.so (offset 0x664000) crash. When I researched about this, it seems like it is related to WebViews and disabling Android System Webview. I dont use these in my applications so I was wondering this could be issue about the SDK?

**Comment by thomasmso** [comment]
@bartutiryakioglu - Thanks for reporting, do you mean you were able to reproduce the crash by disabling Android System WebView?

**Comment by bartutiryakioglu** [comment]
No, this link https://stackoverflow.com/questions/66361472/app-crashed-with-base-apklibmonochrome-so says that disabling can help to resolve this problem. But I couldn't reproduce this error. We could see that error on some of our users. I am not sure what causes this? I am actually asking that does rewarded ads and interstitial ads uses web view that could cause this error?

**Comment by konekwok** [comment]
same here !

**Comment by fdspbsteam** [comment]
libunity.so
JobQueue::ProcessJobs(JobQueue::ThreadInfo*, void*)
same here !

**Comment by vhnunes** [comment]
Same here

**Comment by thomasmso** [comment]
Hey yall - this is an ongoing investigation, and pretty hard to track without a reproducible case. If anyone is able to reproduce it please do enable and send over our plugin's [verbose logs](https://dash.applovin.com/documentation/mediation/unity/getting-started/advanced-settings#enable-verbose-logging).

**Comment by tracer8** [comment]
Still crash like hell on android...

**Comment by marcinbak** [comment]
I'm seeing this crash as well and did some digging.

It is caused by a WebView crashing WebView render process (which is shared between WebViews). Such crash can be caused, for example, by ad JavaScript using a non-supported browser API, or by a device getting low on memory.

If an app has access to WebView prior to loading content, you can set custom WebViewClient with overridden `onRenderProcessGone` callback returning `true` ([docs](https://developer.android.com/reference/android/webkit/WebViewClient#onRenderProcessGone(android.webkit.WebView,%20android.webkit.RenderProcessGoneDetail))). When returning `true` you have to make sure to remove the WebView from view hierarchy and call destroy on it.
If you are able to do this to all currently created WebViews, then the application won't crash.

Having said that, this is usually not possible to achieve if content is loaded to WebViews that are outside of app's control, like with AppLovin adapters.
Ideally, AppLovin and ad SDKs should override the callback in a custom WebViewClient for all WebViews.

**Comment by cemtuver-applovin** [comment]
Hi @marcinbak! Thank you for reporting this issue. I want to assure you that we implement the `onRenderProcessGone` function in all of our `WebViewClient` implementations, and we return `true`, as you also mentioned. Additionally, we are actively communicating with other ad networks to ensure they follow the same practice.

For further investigation, could you please share the full crash trace, along with all the ad network adapters/SDKs and their versions used in your application?

**Comment by marcinbak** [comment]
@cemtuver-applovin thank you for quick reply and that's good to hear.
However there still must be something missing. Maybe it's about the timing when the WebViewClient is set vs when the ad content gets loaded? Without looking into code it's just a speculation.
Most of the complaints I've seen while investigating are related to either Unity ads or Google Ads if that's of any help.

When it comes to networks/adapters and SDK versions:

- com.applovin:applovin-sdk | 11.11.3
- com.applovin.mediation:inmobi-adapter | 10.7.4.0
- com.applovin.mediation:smaato-adapter | 22.5.1.0
- com.applovin.mediation:fyber-adapter | 8.2.3.1
- com.applovin.mediation:mintegral-adapter | 16.4.51.0
- com.applovin.mediation:bytedance-adapter | 6.1.0.6.0
- com.applovin.mediation:vungle-adapter | 6.12.1.1
- com.applovin.mediation:verve-adapter | 3.0.0.0
- com.applovin.mediation:mobilefuse-adapter | 1.6.2.0
- com.applovin.mediation:unityads-adapter | 4.9.2.1

I can see some use older versions and I'll make sure these get upgraded.

**Comment by cemtuver-applovin** [comment]
@marcinbak thank you for the information! We are not aware of any known WebView-related issues with those adapters and SDKs. However, as you mentioned, some of the versions seem to be outdated. I would first suggest updating your adapters, as the network SDKs may have released improvements that could resolve the crashes you're experiencing.

Also, could you please share the package name of the application experiencing these crashes? I might be able to analyze your app to check for any problematic `WebViewClient` implementations.
